### PR TITLE
Add config_registry module

### LINKED ISSUE #4401: Create module repository
The PR #4207 moves all module configuration files into the config module. This is conceptually not entirely correct. We should create a module registry that is populated at runtime allowing for a runtime-discovery of all loaded modules. This would also allow for each module to have its own configuration headers without interfering with the main config module.

### PR DESCRIPTION
**Pull request by msimberg**
Allows registering module configuration entries through static objects. This is done through the `config_registry` module which *all* modules now depend on.

Adds the function `hpx::config_registry::add_module_config` for registering new configuration options, the struct `hpx::config_entry::add_module_config_helper` to help with registering new options through static objects, and `hpx::config_entry::get_module_configs` for getting all the registered options. Fixes #4401.

I left the old mechanism for exposing the configuration strings from the main library in place as that'll be slowly removed as things move into modules.

### CODE REVIEW

File: cmake/templates/config_defines_entries_for_modules.cpp.in [hunk 1]
```diff
@@ -6,16 +6,10 @@
 
 // Do not edit this file! It has been generated by the cmake configuration step.
 
-#if !defined(HPX_CONFIG_DEFINES_STRINGS_@module_name_uc@_HPP)
-#define HPX_CONFIG_DEFINES_STRINGS_@module_name_uc@_HPP
+#include <hpx/config_registry.hpp>
```
Comment (hkaiser) [inline cmake/templates/config_defines_entries_for_modules.cpp.in:9]: Don't we need header guards for this file (or `#pragma once`)?

Comment (hkaiser) [reply cmake/templates/config_defines_entries_for_modules.cpp.in:9]: Ohh sorry, this is a source file now - please disregard my comment.

File: libs/config_registry/src/config_registry.cpp [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    }    // namespace detail
+
+    std::vector<module_config> const& get_module_configs()
+    {
+        return detail::get_module_configs();
+    }
+
+    void add_module_config(module_config const& config)
+    {
+        std::cout << "[before] config registry holds " << detail::get_module_configs().size() << " configs" << std::endl;
```
Comment (hkaiser) [inline libs/config_registry/src/config_registry.cpp:30]: Should these debug messages go away?

File: libs/version/src/version.cpp [hunk 3]
```diff
@@ -199,15 +200,17 @@ namespace hpx {
         strm << "\n";
 
         // print module configurations
-        module_config_strings const* ms = hpx::config_strings_modules;
-        while (ms->name)
+        auto configs = hpx::config_registry::get_module_configs();
```
Comment (hkaiser) [inline libs/version/src/version.cpp:203]: Should this be `auto const& configs = …` instead?

File: libs/config_registry/src/config_registry.cpp (hunk 2 above)
Comment (msimberg) [reply libs/config_registry/src/config_registry.cpp:30]: They should...

File: libs/version/src/version.cpp (hunk 3 above)
Comment (msimberg) [reply libs/version/src/version.cpp:203]: Yep, thanks!

**Comment by msimberg** [comment]
This should also be inspect- and format-clean now.

**Pull request merged**